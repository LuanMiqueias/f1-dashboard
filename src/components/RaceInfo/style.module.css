.container{
  width: 100%;
  height: 100%;
  display: flex;
  align-items: stretch;
  justify-content: flex-start;
  flex-wrap: wrap;
  gap:8px;
  animation: animaShow .2s ease-out forwards;
  background: var(--gray-light);
  border:solid 2px #eee;
  border-radius: 5px;
}
.content{
  width: 100%;
  display: flex;
  flex-direction: column;
}
.content > p{
  width: 100%;
  height: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
  color:var(--text);
  text-align: center;
  padding:0 1rem
}
@keyframes animaShow {
  0%{
    opacity: 0;
  }

  100%{
    opacity: initial;
  }
}
.imgRace{
  width: 100%;
  height: 260px;
  display: flex;
  align-items: center;
  justify-content: center;
  padding: 1rem;
  background: var(--white);
  border-radius: 5px;
  border-bottom: 3px solid var(--primary-color);
}
.imgRace img{
  width: 100%;
  height: 100%;
  object-fit: contain;
}
.info{
  padding: 1rem;
  flex: 1;
  display: flex;
  align-items: flex-start;
  justify-content: center;
  flex-direction: column;
  gap:8px
}
.info h2{
  font-size: 0.88rem;
  font-weight: 400;
  color:var(--sub-title);
  display: flex;
  align-items: baseline;
  gap:8px
}
.info h2 span{
  color:var(--text);
}