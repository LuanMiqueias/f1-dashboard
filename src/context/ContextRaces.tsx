import React, { ReactNode } from "react";
import api from "../services/api";

interface PropsRaces {
  id: number;
  distance?: string;
  date: string;
  status: string;
  season: number;
  type: string;
  laps: {
    total?: number;
  };
  circuit: {
    id: number;
    name: string;
    image: string;
  };
}
interface IData {
  data: PropsRaces[];
}
interface RacesContent {
  races: IData;
  idCompetition: number;
  oneRace: PropsRaces;
  loading: boolean;
  error: string;
  selectRace: (id: number) => void;
  updateRaces: (data: IData) => void;
  updateIdCompetition: (id: number) => void;
}
export const RacesContext = React.createContext({} as RacesContent);

export const RacesProvider: React.FC<ReactNode> = ({ children }) => {
  const [races, setRaces] = React.useState(null as IData);
  const [idCompetition, setIdCompetition] = React.useState(null as number);
  const [oneRace, setOneRace] = React.useState(null as PropsRaces);
  const [loading, setLoading] = React.useState(true as boolean);
  const [error, setError] = React.useState("" as string);

  React.useEffect(() => {
    //DeverÃ¡ set feito um Fetch para api usando o ID da competicao
    if (idCompetition) {
      setLoading(true);
      (async () => {
        const response = await api.get(
          `races?competition=${idCompetition}&season=2021`,
          {
            headers: {
              "x-rapidapi-host": "v1.formula-1.api-sports.io",
              "x-rapidapi-key": process.env.NEXT_PUBLIC_API_KEY,
            },
          }
        );
        if (response.data.errors.request) {
          setError(
            "Sorry for the inconvenience, this site uses a free API plan and you or someone else has exceeded the limit ðŸ˜­, please try again later."
          );
        }
        setRaces({ data: response.data.response });
        setLoading(false);
      })();
    }
  }, [idCompetition]);

  function updateRaces(data: IData) {
    setRaces(data);
    setOneRace(null);
  }
  function updateIdCompetition(id: number) {
    setIdCompetition(id);
    setOneRace(null);
  }
  function selectRace(id: number) {
    if (races) {
      const race = races.data.filter((item) => {
        return item.id === id;
      });
      setOneRace(race[0]);
    }
  }
  return (
    <RacesContext.Provider
      value={{
        races,
        idCompetition,
        oneRace,
        loading,
        error,
        selectRace,
        updateRaces,
        updateIdCompetition,
      }}
    >
      {children}
    </RacesContext.Provider>
  );
};
