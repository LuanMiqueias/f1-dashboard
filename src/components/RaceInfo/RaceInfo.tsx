import React from "react";
import { RacesContext } from "../../context/ContextRaces";
import styles from "./style.module.css";

const RaceInfo = () => {
  const { oneRace } = React.useContext(RacesContext);

  return (
    <div className={styles.container}>
      {oneRace ? (
        <div className={styles.content}>
          <div className={styles.imgRace}>
            <img
              src={oneRace.circuit.image}
              width="100%"
              alt={oneRace.circuit.name}
            />
          </div>
          <div className={styles.info}>
            <h2>
              Status: <span>{oneRace.status}</span>
            </h2>
            <h2>
              Distance:
              <span>
                {oneRace.distance ? oneRace.distance : "no informations"}
              </span>
            </h2>
            <h2>
              Laps:
              <span>
                {oneRace.laps.total ? oneRace.laps.total : "no informations"}
              </span>
            </h2>
          </div>
        </div>
      ) : (
        <div className={styles.content} style={{ minHeight: "400px" }}>
          <p className={styles.text}>Select one race for more informations</p>
        </div>
      )}
    </div>
  );
};

export default RaceInfo;
