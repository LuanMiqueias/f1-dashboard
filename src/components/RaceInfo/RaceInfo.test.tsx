import { act, render, screen } from "@testing-library/react";
import { RacesContext, RacesProvider } from "../../context/ContextRaces";
import RaceInfo from "./RaceInfo";

const racesFake = {
  data: [
    {
      id: 1,
      date: "1995-12-17T03:24:00",
      status: "string",
      season: 1,
      type: "string",
      laps: {
        total: 10,
      },
      circuit: {
        id: 1,
        name: "string",
        image: "string",
      },
    },
  ],
};

it("Should math snapshot", () => {
  const contextCallback = jest.fn();

  const { container } = render(
    <RacesProvider>
      <RaceInfo />
      <RacesContext.Consumer>{contextCallback}</RacesContext.Consumer>
    </RacesProvider>
  );
  act(() => {
    contextCallback.mock.calls[0][0].updateRaces(racesFake);
  });
  act(() => {
    contextCallback.mock.calls[1][0].selectRace(1);
  });

  expect(container.firstChild).toMatchSnapshot();
});

it("Should show the message if oneRace is equal to null", () => {
  render(<RaceInfo />);
  const text = screen.getByText(/Select one race for more informations/i);
  expect(text).toBeInTheDocument();
});
