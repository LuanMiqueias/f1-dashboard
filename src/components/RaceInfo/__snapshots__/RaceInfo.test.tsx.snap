// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should math snapshot 1`] = `
<div
  class="container"
>
  <div
    class="content"
    style=""
  >
    <div
      class="imgRace"
    >
      <img
        alt="string"
        src="string"
        width="100%"
      />
    </div>
    <div
      class="info"
    >
      <h1>
        string
      </h1>
      <h2>
        Type: 
        <span>
          string
        </span>
      </h2>
      <h2>
        Status: 
        <span>
          string
        </span>
      </h2>
      <h2>
        Distance:
        <span>
          no informations
        </span>
      </h2>
      <h2>
        Laps:
        <span>
          10
        </span>
      </h2>
    </div>
  </div>
</div>
`;
